[MRG + 1] Clarified error msg in plot_partial_dependence

kgilliam125 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->
#7600
#### What does this implement/fix? Explain your changes.

Clarified error message in plot_partial_dependence when an
index greater than the number of features is specified in 
`features`
#### Any other comments?

This PR picks up on a previously abandoned PR, #7609 .

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

kgilliam125 committed:
Clarified error msg in plot_partial_dependence

nelson-liu commented on sklearn/ensemble/partial_dependence.py:
```diff
@@ -306,8 +306,8 @@ def convert_feature(fx):
-        raise ValueError('features[i] must be in [0, n_features) '
-                         'but was %d' % i)
+        raise ValueError('features[i] must be in [0, len(feature_names)={0} ) '
+                         'but was {1}'.format(len(feature_names), i))
```
I find it a bit odd that you specify that "features[i] must be in a range", but then show the user the value of "i" in the error message.

kgilliam125 replied:
It was a bit of a tricky issue here; perhaps the error message needs further revision. 

I can't show `features[i]` in this case, since `i` is out of range of the array. Showing the value for `len(feature_names` is also a bit confusing, since both seem like they should contain at most `n_features-1` elements.

From what I can tell based on the documentation, `features` is an array of indices for the features whose names correspond to the elements of `feature_names` at `features[i]`. My goal here is to work out a way to show a message such that a user who hasn't fully read the documentation would understand it.

nelson-liu replied:
ah I see. perhaps @tetraptych would have an opinion about this?

kgilliam125 replied:
There is a bit of history here that's missing, because I picked this up from an abandoned PR. The error message that I have here is, more or less, the same as that proposed in the previous PR, #7609 .

nelson-liu replied:
thanks for linking that to me. yeah, I agree that the error message is mostly there. It just seemed extremely ambiguous to me when I first read it (which made me question its value to a user), but it's obvious in hindsight after I've seen the code in question.

amueller replied:
It's slightly confusing that the ``i`` in ``features[i]`` is not ``i`` as far as I can tell.

amueller replied:
Why not ``features contains {} which is too large for feature_names``? or something like that?
Maybe ``"All entries of features must be <len(feature_names)={}, got {}".format(len(feature_names), features[i])"``

kgilliam125 replied:
Unless I'm missing something, after the `features` is converted by the prior code, the looping below causes  `i` to be a value from the original `features` array that was passed in (if `features` containing integer values).

```
for fxs in features:
            l = []
            # explicit loop so "i" is bound for exception below
            for i in fxs:
                l.append(feature_names[i])
            names.append(l)
    except IndexError:
    ...
```

Part of the issue may come from reassigning the `features` parameter that was passed in to `tmp_features`. After that, any reference to `features[i]` in the code takes a different meaning than what a user thinks of when reading `features[i]`.

amueller commented:
Can you please fix the documentation for ``features`` to say that it can also include strings that are feature names?

kgilliam125 committed:
Changed err msg for feature[i] out of range. Updated docs.

kgilliam125 commented:
@amueller I updated the error message with your suggestion, but left off `features[i]` I can't use it in this case because `i` won't always be in the range [0 len(features)) either if both `features` and `feature_names` are specified. 

I updated the docs to cover all of the ways to specify `features`, so hopefully that at least helps clear up confusion for future use.

amueller commented:
Sorry it should have been ``"got {}".format(i)`` ``i`` was definitely contained in ``features`` which is what was passed.

kgilliam125 committed:
Error message shows invalid value.

amueller commented:
Thanks, lgtm!

NelleV commented:
Thanks for the patch!
